File: project-euler/167/euler-167-v1.pl
#!/usr/bin/perl

use strict;
use warnings;

use integer;
use bytes;

use Tree::RB;

# use Math::BigInt lib => 'GMP', ':constant';

use List::Util qw(sum);
use List::MoreUtils qw();

STDOUT->autoflush(1);

my $DESIRED_K = int( 1 . ( '0' x 11 ) ) - 1;

sub solve
{
    my ($n) = @_;

    my $AA = 2;
    my $BB = $n * 2 + 1;

    my $U = 0;
    my $u = '';
    vec( $u, $U++, 32 ) = $AA;
    vec( $u, $U++, 32 ) = $BB;

    print "$AA\n$BB\n";

    my $v = '';
    vec( $v, $AA + $BB, 2 ) = 1;

    my $CALC_CNT = (
          ( $n < 5 )  ? ( $n * 2_000 )
        : ( $n < 8 )  ? ( $n * 16_000 )
        : ( $n == 8 ) ? ( $n * 500_000 )
        : ( $n == 9 ) ? ( $n * 1_000_000 )
        :               ( $n * 4_000_000 )
    );
    my $FIND_NUM = ( $CALC_CNT >> 2 );
    my $MAX_STEP = ( $FIND_NUM >> 2 );

    my $double_good_next;
    my $even;

    my $next;
    while ( not( defined($double_good_next) and $next > $double_good_next ) )
    {
        $next = undef;
        my $LAST_u = vec( $u, $U - 1, 32 );
    FIND_NEXT:
        for my $i ( $LAST_u + 1 .. ( $LAST_u << 1 ) )
        {
            if ( vec( $v, $i, 2 ) == 1 )
            {
                $next = $i;
                last FIND_NEXT;
            }
        }

        if ( !defined($next) )
        {
            die "Next not found!";
        }

        for my $prev_i ( 0 .. $U - 1 )
        {
            my $s = vec( $u, $prev_i, 32 ) + $next;
            if ( vec( $v, $s, 2 ) < 2 )
            {
                vec( $v, $s, 2 )++;
            }
        }

        # print "$next\n";
        vec( $u, $U++, 32 ) = $next;

        if ( not( $next & 0x1 ) )
        {
            $double_good_next = ( ( $even = $next ) << 1 );
        }
    }

    while ( $U < $CALC_CNT )
    {
        my $val = sub {

            # Pivot for 2
            my $p2 = $U - 1;

            # Pivot for even.
            my $pe = $p2 - 1;

            my $LAST_u = vec( $u, $p2, 32 );

            my $val     = $LAST_u + $AA;
            my $MAX_VAL = $LAST_u + $even;

            while ( vec( $u, $pe, 32 ) + $even > $val )
            {
                $pe--;
            }
            if ( vec( $u, $pe, 32 ) + $even != $val )
            {
                return $val;
            }
            while ( ( $val += 2 ) < $MAX_VAL )
            {
                while ( vec( $u, $pe, 32 ) + $even < $val )
                {
                    $pe++;
                }
                if ( vec( $u, $pe, 32 ) + $even == $val )
                {
                    return $val;
                }
            }
            return $val;
            }
            ->();
        vec( $u, $U++, 32 ) = $val;
    }

    my $M = $U - 1;
STEP:
    for my $STEP ( 1 .. $MAX_STEP )
    {
        my $delta = vec( $u, $M, 32 ) - vec( $u, $M - $STEP, 32 );

        for my $i ( $M - $FIND_NUM .. $M - $STEP - 1 )
        {
            if ( vec( $u, $i + $STEP, 32 ) - vec( $u, $i, 32 ) != $delta )
            {
                next STEP;
            }
        }
        my $STEP_K = $DESIRED_K;
        $STEP_K -= ( ( int( ( $STEP_K - $M ) / $STEP ) + 1 ) * $STEP );
        return
            vec( $u, $STEP_K, 32 ) +
            $delta * ( ( $DESIRED_K - $STEP_K ) / $STEP );
    }

    die "Cannot find for $n!";
}

my $total_sum = 0;
for my $n ( 2 .. 10 )
{
    $total_sum += solve($n);
    print "Total Sum [n=$n] = $total_sum\n";
}
print "Total Sum = $total_sum\n";

=head1 COPYRIGHT & LICENSE

Copyright 2017 by Shlomi Fish

This program is distributed under the MIT / Expat License:
L<http://www.opensource.org/licenses/mit-license.php>

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

=cut

File: project-euler/167/euler_167_v1.cpp
// The Expat License
//
// Copyright (c) 2017, Shlomi Fish
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

#include <vector>
#include <stdlib.h>
#include <string.h>
#include <iostream>

typedef long long ll;

const long GROW_BY = 1024*16;

const ll DESIRED_K = (100000000000LL - 1);

ll solve(unsigned int n)
{
    std::cout << "Inspecting " << n << std::endl;
    const int AA = 2;
    const int BB = n*2+1;

    const unsigned int CALC_CNT = ((n < 8) ? (1000 << n) : (4000 << n));
    const unsigned int FIND_NUM = (CALC_CNT>>2);
    const unsigned int MAX_STEP = FIND_NUM/2;

    std::vector<int> u;

    u.push_back(AA);
    u.push_back(BB);
    // print "AA\nBB\n";

    long v_size = GROW_BY;
    unsigned char * v = (unsigned char *)malloc(v_size+1);
    memset(v, '\0', v_size+1);
    v[AA+BB] = 1;

    std::cout << "[] " << AA << std::endl;
    std::cout << "[] " << BB << std::endl;
    while (u.size() < CALC_CNT)
    {
        int next = -1;
        const int last_elem = u.back();
        const int double_last_elem = last_elem<<1;
        for (int i=last_elem+1; i < double_last_elem; i++)
        {
            if (v[i] == 1)
            {
                next = i;
                break;
            }
        }

        if (next < 0)
        {
            std::cerr << "Next not found!" << std::endl;
            exit(-1);
        }

        const long max_size = next+last_elem;
        if (max_size > v_size)
        {
            const long new_v_size = max_size + GROW_BY;
            auto new_v = (unsigned char *)realloc(v, new_v_size+1);
            if (! new_v)
            {
                free (v);
                return -1;
            }
            v = new_v;
            memset(v+v_size+1, '\0',new_v_size-v_size);
            v_size = new_v_size;
        }
        for (int & prev : u)
        {
            int s = prev+next;
            if (v[s] < 2)
            {
                v[s]++;
            }
        }

        std::cout << '[' << (next - u.back()) << "] " << next << std::endl;
        u.push_back(next);
    }

    free (v);
    v = NULL;

    const int M = u.size()-1;
    for (unsigned int STEP=1; STEP < MAX_STEP; STEP++)
    {
        const int delta = u[M]-u[M-STEP];

        for (unsigned int i = M-FIND_NUM ; i < M-STEP; i++)
        {
            if (u[i+STEP] - u[i] != delta)
            {
                goto end_of_STEP;
            }
        }
        {
            ll STEP_K = DESIRED_K;
            STEP_K -= ((((STEP_K - M) / STEP) + 1) * STEP);
            return u[STEP_K] + delta * ((DESIRED_K - STEP_K) / STEP);
        }

end_of_STEP:
        ;
    }

    std::cerr << "Cannot find for " << n << "!" << std::endl;
    exit(-1);
    return -1;
}

int main()
{
    ll total_sum = 0;
    for (int n = 2 ; n <= 10; n++)
    {
        total_sum += solve(n);
    }
    std::cout << "Total Sum = " << total_sum << std::endl;
    return 0;
}
